File: csharp/1146.snapshot-array.cs
/*
 * @lc app=leetcode id=1146 lang=csharp
 *
 * [1146] Snapshot Array
 */

using System.Collections.Generic;

// @lc code=start
public class SnapshotArray {

    private int Snapshots;
    private IList<(int Snapshot, int Value)>[] Array;

    public SnapshotArray(int length) {
        Snapshots = 0;
        Array = new IList<(int Snapshot, int Value)>[length];
    }
    
    public void Set(int index, int val) {
        var list = Array[index];
        if (list == null) {
            list = new List<(int Snapshot, int Value)>();
            Array[index] = list;
        }
        if (list.Count > 0) {
            var last = list[list.Count - 1];
            if (last.Value == val) {
                return;
            }
            if (last.Snapshot == Snapshots) {
                list.RemoveAt(list.Count - 1);
            }
        }
        list.Add((Snapshot: Snapshots, Value: val));
    }
    
    public int Snap() {
        return Snapshots++;
    }
    
    public int Get(int index, int snap_id) {
        var list = Array[index];
        if (list == null) {
            return 0;
        }
        int l = 0, r = list.Count - 1;
        while (l < r - 1) {
            int m = l + (r - l) / 2;
            if (list[m].Snapshot > snap_id) {
                r = m - 1;
            } else {
                l = m;
            }
        }
        while (r >= 0 && list[r].Snapshot > snap_id) {
            --r;
        }
        if (r < 0) {
            return 0;
        }
        return list[r].Value;
    }
}

/**
 * Your SnapshotArray object will be instantiated and called as such:
 * SnapshotArray obj = new SnapshotArray(length);
 * obj.Set(index,val);
 * int param_2 = obj.Snap();
 * int param_3 = obj.Get(index,snap_id);
 */
// @lc code=end

File: java/1146.snapshot-array.java
/*
 * @lc app=leetcode id=1146 lang=java
 *
 * [1146] Snapshot Array
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// @lc code=start
class SnapshotArray {

    private int version = 0;
    private final Map<Integer, List<Snapshot>> indexToSnapshots;

    public SnapshotArray(int length) {
        indexToSnapshots = new HashMap<>();
    }
    
    public void set(int index, int val) {
        var snapshots = indexToSnapshots.computeIfAbsent(index, ignored -> new ArrayList<>());
        if (!snapshots.isEmpty() && snapshots.getLast().version() == version) {
            snapshots.removeLast();
        }
        snapshots.add(new Snapshot(version, val));
    }
    
    public int snap() {
        return version++;
    }
    
    public int get(int index, int snapId) {
        var snapshots = indexToSnapshots.get(index);
        if (snapshots == null) {
            return 0;
        }
        int l = 0, r = snapshots.size() - 1;
        while (l < r - 1) {
            int m = (l + r) / 2;
            if (snapId >= snapshots.get(m).version()) {
                l = m;
            } else {
                r = m - 1;
            }
        }
        while (r >= 0 && snapshots.get(r).version() > snapId) {
            --r;
        }
        if (r < 0) {
            return 0;
        }
        return snapshots.get(r).value();
    }

    private record Snapshot(int version, int value) {}
}

/**
 * Your SnapshotArray object will be instantiated and called as such:
 * SnapshotArray obj = new SnapshotArray(length);
 * obj.set(index,val);
 * int param_2 = obj.snap();
 * int param_3 = obj.get(index,snap_id);
 */
// @lc code=end

